### ISSUE TITLE
Support for multiple compile_commands.json from external projects

### ISSUE THREAD

**Issue by WimYedema** [microsoft/vscode-cmake-tools#915]
## Brief Issue Summary

Support multiple compile_commands.json file in the build directory. This may happen when there are a number of external projects.

### Expected:

The CMake Tools extension scans the build directory for ALL compile_commands.json files, including those from external projects.

### Apparent Behavior:

The extension only checks for a single compile_commands.json in the build directory itself.

### Platform and Versions

- **Operating System**: Windows
- **CMake Version**: 3.12.2
- **VSCode Version**: 1.40.1
- **CMake Tools Extension Version**: 1.2.2
- **Compiler/Toolchain**: gcc 7.1, arm-none-eabi-gcc 7.3.1 20180622

**Comment by bobbrow** [comment]
Can you provide more information about what your workspace looks like?  Are there separate CMake projects not tied together by a single root CMakeLists.txt?  We are still working on multi-root support, so this is something we can take into consideration, but we need to understand the usage case better before we can deliver this feature.

Thanks!

**Comment by thomthom** [comment]
I'm not sure if this is exactly what the OP had in mind, but I set up my project to use different build directory per config:

```json
{
  "cmake.sourceDirectory": "${workspaceRoot}/cext",
  "cmake.buildDirectory": "${workspaceRoot}/cext/build/${buildType}"
}
```

I do this because I don't like rebuilding completely when I switch between release/debug.

Problem then arise when I in `c_cpp_properties` need to pick one of the `compile_commands.json`:
```json
{
  "compileCommands": "${workspaceFolder}/cext/build/Release/compile_commands.json"
}
```
 This would force me to have to update the JSON file when I switch config.

I'm used to build with Visual Studio, and I like the separation of directories for different configs. When using CMake directly I've usually created a separate build directory per config. But I'm not finding an ideal setup for this with CMake Tools.

**Comment by bobbrow** [comment]
@thomthom, though perhaps a bit inconvenient, in `c_cpp_properties.json` you can define multiple configs and have each one point to a different compileCommands:

```
"configurations": [
  {
    "name": "Debug",
    "compileCommands": "${workspaceFolder}/cext/build/Debug/compile_commands.json"
  },
  {
    "name": "Release",
    "compileCommands": "${workspaceFolder}/cext/build/Release/compile_commands.json"
  }
]
```

Unfortunately, you'd have to remember to change the cpptools config each time you change the cmake config, but this should give you what you want if you need to use compileCommands instead of assigning CMake Tools as the "configurationProvider" (which is the recommended approach).

**Comment by thomthom** [comment]
Sounds like this relates directly to my separate question #928 (I'll continue there.)

**Comment by bobbrow** [comment]
I'm sorry, but we don't have enough information to proceed with this feature request. (@thomthom's issue was addressed in #928)

**Comment by giellamoswhard** [comment]
@bobbrow 
In my case I develop the embedded software for an arm zephyr board and its x86 Linux drivers so I have two compilation databases..

**Comment by sslupsky** [comment]
@bobbrow I develop embedded software for Zephyr RTOS as well.  The compile_commands.json works great for the project directory.  However, it doesn't work so well for the underlying RTOS directory.  I get errors such as:
```
"i2c_sam0.c" not found in "/Users/.../build/compile_commands.json". 'includePath' from c_cpp_properties.json will be used for this file instead.
```
This particular file (i2c_sam0.c) is a Zephyr driver and not part of the application.  When working with your application, it is often helpful to view the drivers used by the application (and other components of the RTOS such as kernel, sub systems, etc) and to have intellisense properly parse these files when you open them.  Moreover, I debug these drivers and it is essential to be able to properly parse them.

**Comment by bobbrow** [comment]
@giellamoswhard, @sslupsky, I created a new issue to discuss your case #1120.  Let's continue the discussion there.

**Comment by nonchip** [comment]
could this be reopened? having the same exact issue, and while OP didn't explain exactly what they meant when asked, i feel this is still valid for this ticket: the use case *is* specified, if you read "external projects" as "calls to `ExternalProject_Add`".

this also isn't a multi-workspace/root issue.

example:
- /CMakeLists.txt `ExternalProject_Add`s /subproject/CMakeLists.txt (or anything else really, even remote git repos, ...)
- when *configured*, `build/compile_commands.json` will contain the commands specified in /CMakeLists.txt *itself*.
- when *built*, additional `compile_commands.json` files will be generated in the external project's `BINARY_DIR`.

if parsing the cmakelists is too complex for this (especially since often `ExternalProject_Add` is called by included functions), i'd suggest simply allowing to specify more than one `compile_commands.json` file.

I tried to work around this limitation by using an abomination like this:
```cmake
ExternalProject_Add(foo ...)
ExternalProject_Add_Step(foo compile_commands
	COMMENT "Fetching compile_commands for foo"
	COMMAND jq -rs ".[0] + .[1]" "${CMAKE_BINARY_DIR}/compile_commands.json" "${CMAKE_BINARY_DIR}/standalone/compile_commands.json" > "${CMAKE_BINARY_DIR}/compile_commands.json"
	DEPENDEES install
	ALWAYS ON # this will probably mess up any build-only-when-needed logic
)
```
but that actually seems to empty `"${CMAKE_BINARY_DIR}/compile_commands.json"` before reading it (despite `-s` on jq being specifically meant to prevent this), so it works but *only* for the external projects.

**Comment by nonchip** [comment]
@bobbrow thanks, that works perfectly fine!
